testing/src/metrics/mod.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    network::{handshaken_peer, test_node, FakeNode, TestSetup},
+    wait_until,
+};
+use snarkos_metrics::stats::NODE_STATS;
+use snarkos_network::Version;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpListener,
+};
+
+#[tokio::test]
+async fn connect_and_disconnect_responder_side() {
+    let setup: TestSetup = Default::default();
+    let node = test_node(setup).await;
+    node.initialize_metrics().await.unwrap();
+
+    // The fake node connects to the node's listener...
+    let peer = handshaken_peer(node.expect_local_addr()).await;
+
+    // Needed to make sure the values have been updated.
+    wait_until!(5, node.peer_book.get_connected_peer_count() == 1);
+
+    // ...the metrics should reflect this.
+    let metrics = NODE_STATS.snapshot();
+
+    assert_eq!(metrics.connections.all_accepted, 1);
+    assert_eq!(metrics.connections.connected_peers, 1);
```

Comment:
I'm missing the `all_initiated`, `disconnected` and `connecting` checks here (not as important, but no harm in being extra thorough); the same applies to the post-disconnect checks and the other test